### PR TITLE
Add embedded PNG texture modifier

### PR DESCRIPTION
**Pull request by hecktest**
This PR builds upon #11485 by adding a texture modifier called `[png:` which allows you to send ephemeral textures to clients in the form of base64 PNG binaries. You can use it for captchas, signs and all sorts of other procedural fun.

mandels

## How to test

Follow the instructions in doc/lua_api.txt, or check out the new devtest nodes:
```
testnodes:generated_png_emb
testnodes:generated_png_src_emb
testnodes:generated_png_dst_emb
```

Ready for review.

### CODE REVIEW

File: src/client/tile.cpp [hunk 1]
```diff
@@ -82,11 +83,9 @@ std::string getImagePath(std::string path)
 {
 	// A NULL-ended list of possible image extensions
 	const char *extensions[] = {
-		"png", "jpg", "bmp", "tga",
-		"pcx", "ppm", "psd", "wal", "rgb",
-		NULL
+		"png", "jpg", "bmp", NULL
```
Comment (hecktest) [inline src/client/tile.cpp:86]: And this is a little opportunistic edit zapping unsupported formats that we're no longer capable of loading.

**Comment by celeron55** [comment]
Looks good

**Comment by hecktest** [comment]
@celeron55 Then approve :o

**Comment by hecktest** [comment]
Reauthored correctly as requested.

**Comment by Emojigit** [comment]
Looks good! afaik that allows custom player texture since this does not require new images in the filesystem, since Minetest disallows adding extra media during client runtime. Ping @Wuzzy2 , he is the MCL2 main contributor, I think this can resolve the custom skin problem of MCL2.

**Comment by Wuzzy2** [comment]
I'm no longer active in MCL2 and other developers took over but nice try. :P

**Comment by Emojigit** [comment]
Will this requires client-side changes? If yes, how to get the client's version?

File: doc/lua_api.txt [hunk 2]
```diff
@@ -620,6 +620,21 @@ Result is more like what you'd expect if you put a color on top of another
...
+`minetest.encode_base64(minetest.encode_png(tex))`,
+refer to the documentation of these functions for details.
+You can use this to send disposable images such as captchas
+to individual clients, or render things that would be too
+expensive to compose with `[combine:`.
+Avoid sending large images in this way. This is not a replacement
+for asset files, do not use it to do anything that you could instead
+achieve by just using a file. Test whether using other
+texture modifiers could result in a shorter string than
+embedding a whole image, this may vary by use case.
```
Comment (sfan5) [inline doc/lua_api.txt:636]: I'd reword this a bit (considering better dynamic media soon)
```suggestion

IMPORTANT: Avoid sending large images this way.
This is not a replacement for asset files, do not use it to do anything
that you could instead achieve by just using a file.
In particular consider `minetest.dynamic_add_media` and test whether
using other texture modifiers could result in a shorter string than
embedding a whole image, this may vary by use case.

```

**Review by sfan5 — APPROVED** [review]
Looks good, CI tests this already.

please look at the review comment though

File: doc/lua_api.txt (hunk 2 above)
Comment (nerzhul) [reply doc/lua_api.txt:636]: haha, we already know that some servers will complain about lag introduced by dynamic png using that feature :D

**Comment by kilbith** [comment]
5.5 milestone please.

**Comment by Emojigit** [comment]
What will older clients do when they received a png modifier?

hecks ***@***.***> 於 2021年7月30日週五 上午3:25 寫道：

**Comment by NathanSalapat** [comment]
I would expect the client to show an unknown texture error and use a random color. Much as happens now it you forget to include a texture, or have a typo in the texture name.

**Comment by appgurueu** [comment]
> I would expect the client to show an unknown texture error and use a random color. Much as happens now it you forget to include a texture, or have a typo in the texture name.

Plus a "generateImage()" error, "unknown texture modifier" or something.

File: doc/lua_api.txt (hunk 2 above)
Comment (hecktest) [reply doc/lua_api.txt:636]: I smell a conflict of interest...

Comment (sfan5) [reply doc/lua_api.txt:636]: There sure is but that doesn't make it wrong to point to a feature that satisfies the same usecase.
